File: GeeksforGeeks/C++/Medium/Longest Repeating Subsequence/longest-repeating-subsequence.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
	public:
		int LongestRepeatingSubsequence(string str){
		    // Code here
		    int n = str.size();
	        vector<vector<int>> dp(n+1,vector<int>(n+1,0));
	        for(int i = 1;i<=n;i++)
	        for(int j = i;j<=n;j++)
	            if(str[i-1]==str[j-1] && i!=j) dp[i][j] = dp[i-1][j-1]+1;
	            else dp[i][j] = max(dp[i-1][j],dp[i][j-1]);
	        return dp[n][n];
		}

};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		string str;
		cin >> str;
		Solution obj;
		int ans = obj.LongestRepeatingSubsequence(str);
		cout << ans << "\n";
	}
	return 0;
}
// } Driver Code Ends

File: GeeksforGeeks/Python3/Medium/Longest Repeating Subsequence/longest-repeating-subsequence.py
#User function Template for python3

class Solution:
	def LongestRepeatingSubsequence(self, str):
		# Code here
        n = len(str)
        str = '.' + str
        prev = [0] * (n+1)
        curr = [0] * (n+1)
        for i in range(1, n+1):
            for j in range(1, n+1):
                if str[i] == str[j] and i != j:
                    curr[j] = 1 + prev[j-1]
                else:
                    curr[j] = max(prev[j], curr[j-1])
            prev = curr.copy()
        return curr[n]

#{ 
 # Driver Code Starts
#Initial Template for Python 3

if __name__ == '__main__':
	T=int(input())
	for i in range(T):
		str = input()
		ob = Solution()
		ans = ob.LongestRepeatingSubsequence(str)
		print(ans)

# } Driver Code Ends